src/libraries/sortedLinkedList.se

```diff
@@ -0,0 +1,164 @@
+data comparor
+data collection[](exists, prev, next)
+data count
+data head
+data tail
+data owner
+data initialized
+
+inset('../macros/require.sem')
+
+extern comparisonProviderInterface: [compare:[int256, int256]:int256]
+
+def init():
+    self.owner = msg.sender
+
+def any():
+    require(msg.sender == self.owner or msg.sender == self)
+
+def initialize(comparor: address):
+    require(not self.initialized)
+    self.initialized = 1
+    self.comparor = comparor
+    return(1)
+
+def add(item: int256, hints: arr):
+    if (self.contains(item)):
+        return(0)
+
+    # If this is the only node it is the head / tail and we can early out
+    if (self.count == 0):
+        self.head = item
+        self.tail = item
+        self.collection[item].exists = 1
+        self.count += 1
+        return(1)
+
+    # By default we begin the insertion search at the head
+    nodeValue = self.head
+
+    # If hints are provided (and valid) we will use the first still present value to begin the search
+    numHints = len(hints)
+    hintIndex = 0
+    while(hintIndex < numHints and node == self.head):
+        hint = hints[hintIndex]
+        if (self.contains(hint) and self.comparor.compare(hint, item) == 1):
+            nodeValue = hint
+
+    # Do traversal to find the insertion point
+    while(self.hasPrev(nodeValue) and self.comparor.compare(nodeValue, item) == 1):
+        nodeValue = self.getPrev(nodeValue)
+    
+    # Actually insert the node
+    if (self.comparor.compare(nodeValue, item) == 1):
```

Comment:
I _think_ it works as is now but given the confusion resulting from how it's written (and the change in behavior with regard to equal members) I'll reorganize it to make it more clear.